File: app/src/main/java/pl/laskowski/marcin/model/SudokuHintGrid.java
package pl.laskowski.marcin.model;

import pl.laskowski.marcin.solving.deduction.algorithm.HintEliminationAlgorithm;
import pl.laskowski.marcin.type.SudokuVariant;

import java.util.*;

/**
 * Created by Marcin Laskowski.
 */

public class SudokuHintGrid {

    private Map<Point, Set<Integer>> hintMap;

    public static SudokuHintGrid createAndReduce(Sudoku sudoku, SudokuVariant sudokuVariant) {
        SudokuHintGrid grid = new SudokuHintGrid(sudoku, sudokuVariant);
        HintEliminationAlgorithm algorithm = new HintEliminationAlgorithm.Factory()
                .instance(sudokuVariant.divideIntoRegions(sudoku), grid);
        algorithm.solve();
        return grid;
    }

    public SudokuHintGrid(Sudoku sudoku, SudokuVariant sudokuVariant) {
        this.hintMap = new HashMap<>();
        for (Field f : sudoku) {
            if (f.isEmpty()) {
                hintMap.put(f.position(), allValuesSet(sudokuVariant));
            }
        }
    }

    private Set<Integer> allValuesSet(SudokuVariant sudokuVariant) {
        int regionSize = sudokuVariant.regionSize();
        Set<Integer> possibleValues = new HashSet<>();
        for (int i = 1; i <= regionSize; i++) {
            possibleValues.add(i);
        }
        return possibleValues;
    }

    public Set<Integer> forField(Field field) {
        return new HashSet<>(getFor(field));
    }

    public Set<Integer> forRegion(Region region) {
        Set<Integer> set = new HashSet<>();
        for (Field field : region) {
            set.addAll(getFor(field));
        }
        return set;
    }

    public boolean isEmpty(Field field) {
        return getFor(field).isEmpty();
    }

    public void clear(Field field) {
        getFor(field).clear();
    }

    public boolean remove(Field field, int value) {
        return getFor(field).remove(value);
    }

    public boolean remove(Region region, int value) {
        boolean removed = false;
        for (Field field : region) {
            removed |= getFor(field).remove(value);
        }
        return removed;
    }

    public boolean contains(Field field, int value) {
        return getFor(field).contains(value);
    }

    public void replace(Field field, Set<Integer> hints) {
        Set<Integer> set = getFor(field);
        set.clear();
        set.addAll(hints);
    }

    public void removeAll(Field field, Set<Integer> hintsToRemove) {
        getFor(field).removeAll(hintsToRemove);
    }

    private Set<Integer> getFor(Field field) {
        return hintMap.getOrDefault(field.position(), Collections.emptySet());
    }

}

File: app/src/main/java/pl/laskowski/marcin/model/SudokuHintGrid.kt
package pl.laskowski.marcin.model

import pl.laskowski.marcin.type.SudokuVariant
import java.util.HashMap
import java.util.HashSet
import pl.laskowski.marcin.solving.deduction.algorithm.HintEliminationAlgorithm

/**
 * Created by Marcin Laskowski.
 */
class SudokuHintGrid(sudoku: Sudoku, sudokuVariant: SudokuVariant) {
    private val hintMap: MutableMap<Point, MutableSet<Int>>

    init {
        hintMap = HashMap()
        for (f in sudoku) {
            if (f.isEmpty) {
                hintMap[f.position] = allValuesSet(sudokuVariant)
            }
        }
    }

    private fun allValuesSet(sudokuVariant: SudokuVariant): MutableSet<Int> {
        val regionSize = sudokuVariant.regionSize()
        return (1..regionSize).toMutableSet()
    }

    fun forField(field: Field): Set<Int> {
        return HashSet(getFor(field))
    }

    fun forRegion(region: Region): Set<Int> {
        val set: MutableSet<Int> = HashSet()
        for (field in region) {
            set.addAll(getFor(field))
        }
        return set
    }

    fun isEmpty(field: Field): Boolean {
        return getFor(field).isEmpty()
    }

    fun clear(field: Field) {
        getFor(field).clear()
    }

    fun remove(field: Field, value: Int): Boolean {
        return getFor(field).remove(value)
    }

    fun remove(region: Region, value: Int): Boolean {
        var removed = false
        for (field in region) {
            removed = removed or getFor(field).remove(value)
        }
        return removed
    }

    fun contains(field: Field, value: Int): Boolean {
        return getFor(field).contains(value)
    }

    fun replace(field: Field, hints: Set<Int>) {
        val set = getFor(field)
        set.clear()
        set.addAll(hints)
    }

    fun removeAll(field: Field, hintsToRemove: Set<Int>) {
        getFor(field).removeAll(hintsToRemove)
    }

    private fun getFor(field: Field): MutableSet<Int> {
        return hintMap.getOrDefault(field.position(), mutableSetOf())
    }

    companion object {
        @JvmStatic
        fun createAndReduce(sudoku: Sudoku, sudokuVariant: SudokuVariant): SudokuHintGrid {
            val grid = SudokuHintGrid(sudoku, sudokuVariant)
            val algorithm = HintEliminationAlgorithm.Factory()
                .instance(sudokuVariant.divideIntoRegions(sudoku), grid)
            algorithm.solve()
            return grid
        }
    }
}